account_payment_mode_auto_reconcile/models/account_invoice.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+import json
+
+from odoo import api, fields, models, _
+
+
+class AccountInvoice(models.Model):
+
+    _inherit = "account.invoice"
+
+    # Allow changing payment mode in open state
+    # TODO: Check if must be done in account_payment_partner instead
+    payment_mode_id = fields.Many2one(
+        states={'draft': [('readonly', False)], 'open': [('readonly', False)]}
+    )
+    payment_mode_warning = fields.Char(
+        compute='_compute_payment_mode_warning',
+    )
+    display_payment_mode_warning = fields.Boolean(
+        compute='_compute_payment_mode_warning',
+    )
+
+    @api.multi
+    def action_invoice_open(self):
+        res = super(AccountInvoice, self).action_invoice_open()
+        for invoice in self:
+            if invoice.type != 'out_invoice':
+                continue
+            if not invoice.payment_mode_id.auto_reconcile_outstanding_credits:
+                continue
+            partial_allowed = self.payment_mode_id.auto_reconcile_allow_partial
+            invoice.auto_reconcile_credits(partial_allowed=partial_allowed)
+        return res
+
+    @api.multi
+    def write(self, vals):
+        res = super(AccountInvoice, self).write(vals)
```

Comment:
shouldn't  it would be performed before write?

Reply:
We can't modify `vals` to do the reconciliation, so better do it afterwards